LXStudio-IDE/src/main/java/titanicsend/pattern/tom/Pulse.java

```diff
@@ -0,0 +1,125 @@
+        float phase = this.phase.getValuef();
+        for (Map.Entry<String, TEPanelModel> entry : model.panelsById.entrySet()) {
+            LXPoint[][] panelPoints = pointMap.get(entry.getKey());
+            int litIndex = (int) (phase * (panelPoints.length - 1));
+            LXPoint[] litSection = panelPoints[litIndex];
+            for (LXPoint point : litSection) {
+                colors[point.index] = LXColor.WHITE;
+            }
+
+            for (int i = 0; i < panelPoints.length; i++) {
+                if (i == litIndex) continue;
+                for (LXPoint point : panelPoints[i]) {
+                    colors[point.index] = LXColor.BLACK;
+                }
+            }
+        }
+    }
+
+    private HashMap<String, LXPoint[][]> buildPointMap(HashMap<String, TEPanelModel> panels) {
+        return panels.entrySet().stream()
+                .collect(Collectors.toMap(
+                        e -> e.getKey(),
+                        e -> buildPanelMap(e.getValue()),
+                        (a, b) -> a,
+                        HashMap::new
+                ));
+    }
+
+
+    private LXPoint[][] buildPanelMap(TEPanelModel panel) {
+        ArrayList<ArrayList<LXPoint>> points = new ArrayList<ArrayList<LXPoint>>();
+        Set<LXPoint> unassignedPoints = Arrays.stream(panel.points).collect(Collectors.toSet());
+
+        TEVertex[] currentVertices = {
+                new TEVertex(new LXPoint(panel.v0), -1),
+                new TEVertex(new LXPoint(panel.v1), -1),
+                new TEVertex(new LXPoint(panel.v2), -1)
+        };
+
+        int currentIndex = 0;
+        while (currentVertices[0].distanceTo(panel.centroid) > (10 * PanelStriper.DISTANCE_BETWEEN_PIXELS)) {
+            points.add(new ArrayList<LXPoint>());
+            ArrayList<LXPoint> assignedPoints = new ArrayList<LXPoint>();
+            LXPoint[][] edges = {
+                    {currentVertices[0], currentVertices[1]},
+                    {currentVertices[1], currentVertices[2]},
+                    {currentVertices[0], currentVertices[2]},
+            };
+
+            for (LXPoint point : unassignedPoints) {
+                for (LXPoint[] edge : edges) {
+                    if (distanceBetPointAndLine(edge[0], edge[1], point) < 5 * PanelStriper.DISTANCE_BETWEEN_PIXELS) {
+                        points.get(currentIndex).add(point);
+                    }
+                }
+            }
+
+            for (LXPoint point : points.get(currentIndex)) {
+                unassignedPoints.remove(point);
+            }
+
+            for (TEVertex vertex : currentVertices) {
+                vertex.nudgeToward(panel.centroid, (float) 0.05);
+            }
+            currentIndex++;
+        }
+
+        return points.stream()
+                .map(l -> l.stream().toArray(LXPoint[]::new))
+                .toArray(LXPoint[][]::new);
+    }
+
+    private static double distanceBetPointAndLine(LXPoint line_p1, LXPoint line_p2, LXPoint p) {
```

Comment:
Guess we should find a home for this so PanelStriper and this can share it.

Reply:
Yeah, maybe we ought to put together a `GeometryUtil` class or something.